src/main/java/com/vmware/vipclient/i18n/messages/service/ComponentService.java

```diff
@@ -103,48 +100,62 @@ public MessageCacheItem getMessages() {
 	 * 	 	<li>The messages in a fallback locale</li>
 	 * </ul>
 	 */
-    public MessageCacheItem getMessages(Iterator<Locale> fallbackLocalesIter) {
-    	CacheService cacheService = new CacheService(dto);
-    	MessageCacheItem cacheItem = cacheService.getCacheOfComponent();
-    	if (cacheItem != null) { // Item is in cache
-    		if (cacheItem.isExpired()) {
-    			// Refresh the cacheItem in a separate thread
-    			refreshCacheItemTask(cacheItem);
-    		}
-    	} else { // Item is not in cache.
+	public MessageCacheItem getMessages(Iterator<Locale> fallbackLocalesIter) {
+		this.doLocaleMatching();
+
+		CacheService cacheService = new CacheService(dto);
+		MessageCacheItem cacheItem = cacheService.getCacheOfComponent();
+		if (cacheItem != null) { // Item is in cache
+			if (cacheItem.isExpired())
+				refreshCacheItemTask(cacheItem); // Refresh the cacheItem in a separate thread
+		} else { // Item is not in cache.
 			ProductService ps = new ProductService(dto);
-			if (!ps.isSupportedLocale(Locale.forLanguageTag(dto.getLocale()))) { // Requested locale is not supported
-				Locale matchedLocale = LocaleUtility.pickupLocaleFromList(new LinkedList<>(ps.getSupportedLocales()), Locale.forLanguageTag(dto.getLocale()));
-				if (ps.isSupportedLocale(matchedLocale)) { // Requested locale matches a supported locale (eg. requested locale "fr_CA matches supported locale "fr")
-					MessagesDTO matchedLocaleDTO = new MessagesDTO(dto.getComponent(), matchedLocale.toLanguageTag(), dto.getProductID(), dto.getVersion());
-					return new ComponentService(matchedLocaleDTO).getMessages();
-				}
-			}
+			Locale locale = Locale.forLanguageTag(dto.getLocale());
+			if (ps.isSupportedLocale(locale) || VIPCfg.getInstance().isPseudo()) {
+				cacheItem = createCacheItem(); // Fetch for the requested locale from data store, create cacheItem and store in cache
+				if (cacheItem.getCachedData().isEmpty())  // Failed to fetch messages for the requested locale
+					cacheItem = getFallbackLocaleMessages(fallbackLocalesIter);
+			} else   // Requested locale is not supported and does not match any supported locales
+				cacheItem = getFallbackLocaleMessages(fallbackLocalesIter);
+		}
+		return cacheItem;
```

Comment:
Suggest code for this method, then doLocaleMatching can be removed:

		Locale matchedLocale = LocaleUtility.pickupLocaleFromList(new ProductService(dto).getSupportedLocales(),
				Locale.forLanguageTag(dto.getLocale()));
		MessageCacheItem cacheItem = null;
		if (matchedLocale != null) {
			dto.setLocale(matchedLocale.toLanguageTag());
			cacheItem = new CacheService(dto).getCacheOfComponent();
			if (cacheItem != null) { // Item is in cache
				if (cacheItem.isExpired())
					refreshCacheItemTask(cacheItem); // Refresh the cacheItem in a separate thread
			} else { // Item is not in cache.
				cacheItem = createCacheItem(); // Fetch for the requested locale from data store, create cacheItem and store in cache
			}
		}
		
		if (cacheItem == null || cacheItem.getCachedData().isEmpty()) // Failed to fetch messages for the requested locale
			cacheItem = getFallbackLocaleMessages(fallbackLocalesIter);

		return cacheItem;